draft-ietf-ace-coap-est-oscore.md

```diff
@@ -279,6 +279,26 @@ In addition, EST-oscore uses the same CoAP Content-Format identifiers when trans
 |       | application/csrattrs                           (res) |  285  |
 {: #table_mediatypes cols="l l" title="EST functions and the associated CoAP Content-Format identifiers"}
 
+Content-Format 281 MUST be supported by EST-oscore servers.
+Servers MAY also support Content-Format 287.
+It is up to the client to support only Content-Format 281, 287 or both.
+As indicated in {{Section 4.3 of RFC9148}}, the client will use a CoAP Accept Option in the request to express the preferred response Content-Format.
+If an Accept Option is not included in the request, the client is not expressing any preference and the server SHOULD choose format 281.
+
+The generated response for /skg and /skc requests contains two parts: certificate and the corresponding private key.
+{{Section 4.8 of RFC9148}} specifies that the private key in response to /skc request may be either an encrypted (PKCS #7) or unencrypted (PKCS #8) key, depending on whether the CSR request included SMIMECapabilities.
```

Comment:
Basically, Section 4.8 of RFC 9148 (third paragraph, last sentence) is also admitting the combination (284, 287), for responses to both /skg and /skc requests, depending on whether the CSR request included SMIMECapabilities, as you say.

However, Table 3 in Section 4.3 of RFC 9148 gives a different impression. Perhaps the intention was to have "284 or 280" in the column "Response, Part 1" for both rows "/skg" and "/skc"?

Regardless, I think that what's proposed below is a good direction for EST-OSCORE.

Reply:
Hi, many PKCS (embedded) libraries do not speak PKCS7, but only PKCS8.  Meanwhile, many historical non-embedded systems have the opposite requirements and/or the PKCS8 code paths were removed during FIPS evaluations.  That's why RFC9148 decided to let the client decide which one they want.    

Unless EST-OSCORE has a compelling reason to include server generated private keys, I suggest leaving it out completely: it's only asking for SecAD DISCUSSes.   The reason RFC9148 had it was because of very specific situations where an RSA1024-bit IDevID key was to be replaced with an RSA 2048 key, and it was KNOWN that the random number generator in the client machines was poor, and unlikely to be fixed.  Why one vendor's woes had such influence is a different question I can't answer, but let's not repeat it unless we have a really good reason.  I think that any 10yr old silicon has good enough RNG support that it's not an issue.

Reply:
And to be clear, the client says in their Accept: option which format they want.

Reply:
> Unless EST-OSCORE has a compelling reason to include server generated private keys, I suggest leaving it out completely: it's only asking for SecAD DISCUSSes. The reason RFC9148 had it was because of very specific situations where an RSA1024-bit IDevID key was to be replaced with an RSA 2048 key, and it was KNOWN that the random number generator in the client machines was poor, and unlikely to be fixed. Why one vendor's woes had such influence is a different question I can't answer, but let's not repeat it unless we have a really good reason. I think that any 10yr old silicon has good enough RNG support that it's not an issue.

One argument that was brought up by Shahid is that a hardware RNG alone, indeed dominantly present nowadays, is not sufficient to guarantee good quality random numbers. He brought up this talk from DEF CON 2021: https://www.youtube.com/watch?v=Zuqw0-jZh9Y.

As it stands, I think the issue of keeping or removing server-generated private keys deserves more discussion. I will keep it in the draft for now and bring it up for discussion in the ACE interim on 5 June 2023.